sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeHigherOrderFunctions.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Reorders and combines [[HigherOrderFunction]]s on collections such that
+ * the resulting plan would have less nodes and do less operations.
+ */
+object OptimizeHigherOrderFunctions extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformExpressions {
+    case ArrayFilter(ArrayFilter(child, f1: LambdaFunction), f2: LambdaFunction) =>
+      ArrayFilter(child, f1.copy(function = And(f1.function, f2.function)))
```

Comment:
Maybe I am too much careful about this but I think it could cause some behaviour changes in corner cases. For example, let's say `f2.function` throws an exception in some condition. Previously, it threw an exception but now `f1.function` returns `false`, and `f2.function` is not executed by short-circuiting.

Reply:
Maybe we should use `NoThrow`.

Reply:
I think that the filters are executed the same amount of times in both cases.

Reply:
Oops right.